lunacy: init at 9.4.2.5022

eliandoran opened the pull request:
## Description of changes

> (un)Free design software that keeps your flow with AI tools and built-in graphics

Home page: https://icons8.com/lunacy

Closes package request #201926

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

eliandoran committed:
lunacy: init at 9.4.2.5022

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+, libXcursor
+, openssl }:
+
+stdenv.mkDerivation rec {
```
Avoid using `rec` and instead convert this to be a function. See https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+  pname = "lunacy";
+  version = "9.3.3";
+
+  # see https://lunacy.docs.icons8.com/release-notes/
```
Move this comment to a changelog attribute. See https://nixos.org/manual/nixpkgs/stable/#var-meta-changelog

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+    runHook postInstall
+  '';
+
+  postInstall = ''
+    substituteInPlace $out/share/applications/lunacy.desktop \
+      --replace "Exec=/opt/icons8/lunacy/Lunacy" "Exec=lunacy" \
+      --replace "Icon=/opt/icons8/lunacy/Assets/LunacyLogo.png" "Icon=lunacy"
+  '';
```
Consider using `makeDesktopItem` instead. See other packages in nixpkgs for reference. Example: https://github.com/NixOS/nixpkgs/blob/cfc3698c31b1fb9cdcf10f36c9643460264d0ca8/pkgs/games/atlauncher/default.nix#L40-L48

eliandoran replied:
@GetPsyched , although in general it's good to use `makeDesktopItem`, I think we shouldn't use it if there is already a desktop file in the upstream repository. This is because generally desktop files can contain a lot of localized names; although the `lunacy` one does not contain that many items, it defines a mime type and a `StartupWMClass`.

I don't think we should duplicate those fields, instead just fix the icon so that it does not have hard-coded paths.

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+  '';
+
+  postFixup = ''
+    makeWrapper $out/lib/lunacy/Lunacy $out/bin/lunacy \
```
```suggestion
    makeWrapper $out/lib/lunacy/Lunacy $out/bin/${finalAttrs.pname} \
```

doronbehar replied:
@GetPsyched I personally think that this kind of link is useless, so I'd advice against it.

GetPsyched replied:
Can you clarify the reasoning? I want to know why is this bad. Since I was given this suggestion when I pushed my first package too.

doronbehar replied:
It's a debatable subject, see for example:

https://github.com/nix-community/nixpkgs-lint/issues/21

I simply think that this link has no use, and it makes the expression less readable.

GetPsyched replied:
Okay yeah I can see that for the `repo` attribute since overriding the `pname` would break the source pull itself. Though, if I were to override the `pname` as a user, I would personally want the name of the binary to change as well. But yeah, this is also up for debate.

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+    maintainers = with lib.maintainers; [ eliandoran ];
+    platforms = lib.platforms.linux;
+    sourceProvenance = with lib.sourceTypes; [ binaryBytecode ];
+    mainProgram = "lunacy";
```
```suggestion
    mainProgram = finalAttrs.pname;
```

doronbehar replied:
Same here.

SuperSandro2000 replied:
mainProgram should not depend on pname. This was the previous default and we opted out of that.

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,100 @@
+      ]}
+  '';
+
+  meta = {
```
Consider using `with lib;` here so the references inside become less verbose.
```suggestion
  meta = with lib; {
```

eliandoran commented:
@GetPsyched , please re-review.

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+
+    # Prepare the desktop icon, the upstream icon is 200x200 but the hicolor theme does not
+    # support this resolution. Nearest sizes are 192x192 and 256x256.
+    ${imagemagick}/bin/convert "opt/icons8/lunacy/Assets/LunacyLogo.png" -resize 192x192 lunacy.png
```
Is this needed? What is the original resolution? Also, see if an SVG exists for it, it would be preferred

eliandoran replied:
There is only one `.svg` file and that's for one of the file types, unfortunately. We have to downscale the icon because otherwise it will not be displayed as-is (upstream hard-codes the path to the .png instead of putting it into `/usr/share/icons/hicolor`).

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+    # Prepare the desktop icon, the upstream icon is 200x200 but the hicolor theme does not
+    # support this resolution. Nearest sizes are 192x192 and 256x256.
+    ${imagemagick}/bin/convert "opt/icons8/lunacy/Assets/LunacyLogo.png" -resize 192x192 lunacy.png
+    install -D lunacy.png "$out/share/icons/hicolor/192x192/apps/lunacy.png"
```
```suggestion
    install -D lunacy.png "$out/share/icons/hicolor/192x192/apps/${finalAttrs.pname}.png"
```

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+    homepage = "https://icons8.com/lunacy";
+    changelog = "https://lunacy.docs.icons8.com/release-notes/";
+    license = licenses.unfree;
+    maintainers = with maintainers; [ eliandoran ];
```
```suggestion
    maintainers = [ maintainers.eliandoran ];
```

GetPsyched commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+    license = licenses.unfree;
+    maintainers = with maintainers; [ eliandoran ];
+    platforms = platforms.linux;
+    sourceProvenance = with sourceTypes; [ binaryBytecode ];
```
```suggestion
    sourceProvenance = [ sourceTypes.binaryBytecode ];
```

eliandoran commented:
@GetPsyched , please re-review.

GetPsyched commented:
That's it from my end. Can't give an explicit approval since I did not test the build process. Everything else LGTM.

doronbehar reviewed (commented):
Besides the too often `finalAttrs.pname` usage which I think is wrong, these are a few comments on the `patchelf` usage.

GetPsyched commented:
> Besides the too often `finalAttrs.pname` usage which I think is wrong

Yes, I made a mistake for that in `mainProgram`; it should be static there. Ref: https://github.com/NixOS/nixpkgs/pull/275550

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+    cp -R "opt/icons8/lunacy" "$out/lib"
+    cp -R "usr/share" "$out/share"
+
+    patchelf \
+      --add-needed libICE.so.6 \
+      --add-needed libSM.so.6 \
+      --add-needed libX11.so.6 \
+      --add-needed libXcursor.so.1 \
+      "$out/lib/lunacy/Lunacy"
+
+    patchelf \
+      --replace-needed liblttng-ust.so.0 liblttng-ust.so \
+      $out/lib/lunacy/libcoreclrtraceptprovider.so
```
that is need or else autopatchelf errors out with:
```
error: auto-patchelf could not satisfy dependency liblttng-ust.so.0 wanted by /nix/store/5kwswim4wpijv1pv9r93vsya98mzrcga-lunacy-9.3.3/lib/lunacy/libcoreclrtraceptprovider.so
```

eliandoran replied:
@doronbehar , added a comment. This patch is also done manually in other packages such as `ms-vscode.cpptools` VS code extension, `aws-workspaces`

a-n-n-a-l-e-e reviewed (commented):
not sure if this PR is paused and apologize for adding yet another review however -- i added a comment on how to get rid of the LD_LIBRARY_PATH and wrapper.

with those changes the program boots and looking ldd from on ubuntu looks good. when exiting the program it coredumps tho that was happening with the wrapper too.

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+    cp -R "opt/icons8/lunacy" "$out/lib"
+    cp -R "usr/share" "$out/share"
+
+    patchelf \
+      --add-needed libICE.so.6 \
+      --add-needed libSM.so.6 \
+      --add-needed libX11.so.6 \
+      --add-needed libXcursor.so.1 \
+      "$out/lib/lunacy/Lunacy"
```
```suggestion
```
this can be removed and then add:
```nix
  runtimeDependencies = [
    # Runtime deps
    libICE
    libSM
    libX11
    libXcursor
  ];
```
somewhere.

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+
+  dontBuild = true;
+  dontStrip = true;
+
```
```suggestion
  appendRunpaths = map (pkg: (lib.getLib pkg) + "/lib") [ icu openssl stdenv.cc.libc stdenv.cc.cc ]; 
  
  runtimeDependencies = [                                                                                                                             
    # Runtime deps                                                                                                                                    
    libICE                                                                                                                                            
    libSM
    libX11
    libXcursor                                                                                                                                        
  ];           
```
this goes along with my comments about the installPhase and postFixup

eliandoran commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+      --replace "Icon=/opt/icons8/lunacy/Assets/LunacyLogo.png" "Icon=${finalAttrs.pname}"
+  '';
+
+  postFixup = ''
+    makeWrapper $out/lib/lunacy/Lunacy $out/bin/${finalAttrs.pname} \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
+        icu
+        openssl
+      ]}
```
@a-n-n-a-l-e-e , where did you learn all this? :)

a-n-n-a-l-e-e replied:
looking at `pkgs/build-support/setup-hooks/auto-patchelf.sh` and `pkgs/build-support/setup-hooks/auto-patchelf.py`

eliandoran commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,107 @@
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
+        icu
+        openssl
+      ]}
```
@a-n-n-a-l-e-e helped a lot to simplify the code. Please re-review.

eliandoran commented:
@a-n-n-a-l-e-e , I have updated Lunacy to the latest version. Seems the upstream devs did quite a few changes, now the packaging is slightly easier.

However, I still had to use `makeWrapper` because the binary would fail at runtime because it could not find `libSkiaSharp.so` (which we don't have packaged somewhere, but it is provided in the same directory as the executable). So I added the `lib` directory to `LD_LIBRARY_PATH`. Is there any cleaner way to implement this? Maybe by changing the directory structure?

Here's the current tree of `result`:

```
result
├── bin
│   └── lunacy
├── lib
│   └── lunacy
│       ├── Assets
│       │   ├── FileIcon.ico
│       │   ├── Images
│       │   │   └── PatternFills
│       │   │       ├── 049b677d19d023f2dff1acee081f093c781008e3.png
│       │   │       ├── 578dc6ec6ee8247afddacd73b4ec8bbc8edfd8f5.png
│       │   │       ├── 5dd3b2fc9c49af1dbc48c5355bb7000c15ce78cf.png
│       │   │       ├── 88f02b6772be16df3dce5d8d2f227bc83f457532.png
│       │   │       ├── 8fbd768bfbabe02ebe079ab7a09a519fc6bd2bb2.png
│       │   │       └── dc490d9f18a6a859fd6da7b5ec02529909e55db2.png
│       │   ├── Localization.json
│       │   ├── LunacyLogo.png
│       │   ├── PreCache
│       │   │   ├── icons.categories.data
│       │   │   ├── icons.categories.index
│       │   │   ├── icons.data
│       │   │   └── icons.index
│       │   └── Prototyping
│       │       ├── Cloud
│       │       │   ├── menuLeft.html
│       │       │   ├── menuRight.html
│       │       │   └── script.js
│       │       ├── index.html
│       │       ├── Local
│       │       │   ├── menuLeft.html
│       │       │   ├── menuRight.html
│       │       │   └── script.js
│       │       ├── script.js
│       │       ├── Shared
│       │       │   ├── menuLeft.html
│       │       │   ├── menuRight.html
│       │       │   └── script.js
│       │       └── style.css
│       ├── Components
│       │   ├── Address.ldb
│       │   ├── Article.ldb
│       │   ├── Avatars.ldb
│       │   ├── Business.ldb
│       │   ├── Components
│       │   │   └── Default.sketch
│       │   ├── Information
│       │   │   ├── Tutorial.free
│       │   │   └── WhatsNew.free
│       │   ├── Person.ldb
│       │   ├── Technology.ldb
│       │   └── UIKits
│       │       ├── Apple iOS.sketch
│       │       ├── Apple macOS.sketch
│       │       ├── Apple watchOS.sketch
│       │       ├── Google Material.sketch
│       │       ├── Microsoft Web Fluent.sketch
│       │       └── Microsoft Windows Fluent.sketch
│       ├── libe_sqlite3.so
│       ├── libglfw.so.3.3
│       ├── libHarfBuzzSharp.so
│       ├── libSkiaSharp.so
│       └── Lunacy
└── share
    ├── applications
    │   └── lunacy.desktop
    ├── icons
    │   └── hicolor
    │       ├── 192x192
    │       │   └── apps
    │       │       └── lunacy.png
    │       └── scalable
    │           └── mimetypes
    │               └── zip-sketch.svg
    └── mime
        └── packages
```

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+, lib
+, fetchurl
+, dpkg
+, makeWrapper
```
```suggestion
```

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+
+  nativeBuildInputs = [
+    dpkg
+    makeWrapper
```
```suggestion
```

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+    openssl
+    stdenv.cc.libc
+    stdenv.cc.cc
+  ];
```
```suggestion
  ] ++ [
    # technically, this should be in runtimeDependencies but will not work as
    # "lib" is appended to all elements in the array
    "${placeholder "out"}/lib/lunacy"
  ];
```

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+      --replace "Icon=/opt/icons8/lunacy/Assets/LunacyLogo.png" "Icon=${finalAttrs.pname}"
+  '';
+
+  postFixup = ''
+    mkdir $out/bin
+
+    # Fixes runtime error regarding missing libSkiaSharp.so (which is in the same directory as the binary).
+    makeWrapper "$out/lib/lunacy/Lunacy" "$out/bin/${finalAttrs.pname}" \
+      --prefix LD_LIBRARY_PATH : "$out/lib/lunacy"
+  '';
+
```
```suggestion
```

a-n-n-a-l-e-e commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+
+  installPhase = ''
+    runHook preInstall
+
```
```suggestion
    mkdir -p $out/bin
    ln -s $out/lib/lunacy/Lunacy $out/bin

```

a-n-n-a-l-e-e reviewed (commented):
can append `"${placeholder "out"}/lib/lunacy"` to `appendRunpaths` and then remove the makeWrapper. Technically, this could be in `runtimeDependencies` but all items in that array get an extra `lib` appended so will not work.

AndersonTorres commented on pkgs/by-name/lu/lunacy/package.nix:
```diff
@@ -0,0 +1,112 @@
+, libSM
+, libXcursor
+, openssl
+, imagemagick }:
```
```suggestion
, imagemagick 
}:
```

Nit

eliandoran commented:
@a-n-n-a-l-e-e , @AndersonTorres , @doronbehar , please re-review.

a-n-n-a-l-e-e reviewed (approved):
autoPatchelfHook looks good -- not sure what is going on with all the uses of `finallAttrs.pname` but i guess it doesn't really matter.
